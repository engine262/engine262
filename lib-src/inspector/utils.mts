import type Protocol from 'devtools-protocol';
import { Inspector } from './index.mts';
import {
  CreateBuiltinFunction, CreateDataProperty, CreateNonEnumerableDataPropertyOrThrow, DefinePropertyOrThrow, Descriptor, DetachArrayBuffer, isArrayBufferObject, NormalCompletion, OrdinaryObjectCreate, surroundingAgent, ThrowCompletion, skipDebugger, Value, type Arguments, type ManagedRealm,
  type ValueEvaluator,
} from '#self';

const consoleMethods = [
  'log',
  'debug',
  'info',
  'error',
  'warning',
  'dir',
  'dirxml',
  'table',
  'trace',
  'clear',
  'startGroup',
  'startGroupCollapsed',
  'endGroup',
  'assert',
  'profile',
  'profileEnd',
  'count',
  'timeEnd',
] as const;
type ConsoleMethod = typeof consoleMethods[number];
export function createConsole(realm: ManagedRealm, defaultBehaviour: Partial<Record<ConsoleMethod, (args: Arguments) => void | NormalCompletion<void>>>) {
  realm.scope(() => {
    const console = OrdinaryObjectCreate(realm.Intrinsics['%Object.prototype%']);
    skipDebugger(DefinePropertyOrThrow(
      realm.GlobalObject,
      Value('console'),
      Descriptor({
        Configurable: Value.true,
        Enumerable: Value.false,
        Writable: Value.true,
        Value: console,
      }),
    ));
    consoleMethods.forEach((method) => {
      const f = CreateBuiltinFunction(
        (args) => {
          if (surroundingAgent.debugger_isPreviewing) {
            return Value.undefined;
          }
          if (defaultBehaviour[method]) {
            const completion = (defaultBehaviour[method](args));
            if (completion instanceof ThrowCompletion) {
              return completion;
            }
          }
          if (realm.HostDefined.attachingInspector instanceof Inspector) {
            realm.HostDefined.attachingInspector.console(realm, method as Protocol.Protocol.Runtime.ConsoleAPICalledEventType, args);
          }
          return Value.undefined;
        },
        1,
        Value(method),
        [],
      );
      skipDebugger(CreateDataProperty(console, Value(method), f));
    });
  });
}

export function createInternals(realm: ManagedRealm) {
  realm.scope(() => {
    const $ = OrdinaryObjectCreate.from({
      debugger: () => {
        if (surroundingAgent.debugger_isPreviewing) {
          return;
        }
        // eslint-disable-next-line no-debugger
        debugger;
      },
      * detachArrayBuffer(object): ValueEvaluator {
        if (!isArrayBufferObject(object)) {
          return surroundingAgent.Throw('TypeError', 'Raw', 'Argument must be an ArrayBuffer');
        }
        const completion = DetachArrayBuffer(object);
        if (completion instanceof ThrowCompletion) {
          return completion;
        }
        return Value.undefined;
      },
    });
    CreateNonEnumerableDataPropertyOrThrow(realm.GlobalObject, Value('$'), $);
  });
}
