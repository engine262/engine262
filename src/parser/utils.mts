import type { ParseNode } from '#self';

export type TargetSymbol = ParseNode['type'] | 'super' | 'this';

export function Contains(node: ParseNode, symbol: TargetSymbol): boolean {
  switch (node.type) {
    case 'FunctionDeclaration':
    case 'FunctionExpression':
    case 'GeneratorDeclaration':
    case 'GeneratorExpression':
    case 'AsyncGeneratorDeclaration':
    case 'AsyncGeneratorExpression':
    case 'AsyncFunctionDeclaration':
    case 'AsyncFunctionExpression':
      return false;
    case 'ClassTail': {
      // We don't have ClassBody?
      throw new Error('TODO');
    }
    case 'ClassStaticBlock':
      return false;
    case 'ArrowFunction':
    case 'AsyncArrowFunction':
      throw new Error('TODO');
    case 'PropertyDefinition':
      throw new Error('TODO');
    //  LiteralPropertyName : IdentifierName
    // throw new Error('TODO');
    case 'MemberExpression': {
      //  MemberExpression : MemberExpression . IdentifierName
      if (node.IdentifierName) {
        return Contains(node.MemberExpression, symbol);
      }
      break;
    }
    case 'SuperProperty': {
      if (node.IdentifierName) {
        return symbol === 'super';
      }
      break;
    }
    case 'CallExpression': {
      throw new Error('TODO');
    }
    case 'OptionalChain': {
      if (node.IdentifierName) {
        //  OptionalChain : OptionalChain . IdentifierName
        if (node.OptionalChain) {
          return Contains(node.OptionalChain, symbol);
        }
        //  OptionalChain : ?. IdentifierName
        return false;
      }
      break;
    }
    default:
  }

  // 1. For each child node child of this Parse Node
  for (const possibleChildKey in node) {
    if (Object.hasOwn(node, possibleChildKey)) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const possibleChild = (node as any)[possibleChildKey];
      if ('type' in possibleChild) {
        // a. If child is an instance of symbol, return true.
        if (possibleChild.type === symbol) {
          return true;
        }
        // b. If child is an instance of a nonterminal, then
        const contained = Contains(possibleChild, symbol);
        // i. If contained is true, return true.
        if (contained) {
          return true;
        }
      }
    }
  }
  return false;
}
