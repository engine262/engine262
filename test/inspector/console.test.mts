/* eslint-disable no-await-in-loop */
/* eslint-disable quotes */
import { expect, test } from 'vitest';
import { TestInspector } from './utils.mts';
import { Agent, ManagedRealm, setSurroundingAgent } from '#self';

test('compile script (for invalid code break line)', async () => {
  const agent = new Agent();
  setSurroundingAgent(agent);
  const inspector = new TestInspector();
  const realm = new ManagedRealm();
  inspector.attachAgent(agent, [realm]);

  const result = await inspector.runtime.compileScript({
    expression: 'function f() {',
    persistScript: false,
    sourceURL: '',
    executionContextId: 0,
  });
  expect(result).toMatchInlineSnapshot(`
    {
      "exceptionDetails": {
        "columnNumber": 0,
        "exception": {
          "className": "SyntaxError",
          "description": "SyntaxError: Unexpected end of input",
          "objectId": "default:2",
          "preview": {
            "description": "
    function f() {
                  ^
    SyntaxError: Unexpected end of source",
            "entries": undefined,
            "overflow": false,
            "properties": [
              {
                "name": "message",
                "type": "string",
                "value": "Unexpected end of source",
              },
              {
                "name": "stack",
                "type": "accessor",
              },
            ],
            "subtype": "error",
            "type": "object",
          },
          "subtype": "error",
          "type": "object",
        },
        "exceptionId": 1,
        "lineNumber": 0,
        "scriptId": undefined,
        "stackTrace": undefined,
        "text": "Uncaught",
        "url": undefined,
      },
    }
  `);
});

test('preview evaluation', async () => {
  const agent = new Agent();
  setSurroundingAgent(agent);
  const inspector = new TestInspector();
  const realm = new ManagedRealm();
  inspector.attachAgent(agent, [realm]);

  for (const code of [
    `1`,
    `[]`,
    `{ var a = 1; a }`,
    `{ let a = 1; a }`,
    `[1, 2, 3].map(x => x + 1)`,
    `globalThis.x = 1`,
  ]) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const result: any = await inspector.perview(code);
    if (result.exceptionDetails?.exception.preview.properties[0].value === 'Preview evaluator cannot evaluate side-effecting code') {
      expect('side effect').toMatchSnapshot(code);
    } else {
      expect(result).toMatchSnapshot(code);
    }
  }
});

test('get local lexical names', async () => {
  const agent = new Agent();
  setSurroundingAgent(agent);
  const inspector = new TestInspector();
  const realm = new ManagedRealm();
  inspector.attachAgent(agent, [realm]);

  expect(await inspector.runtime.globalLexicalScopeNames({
    executionContextId: 0,
  })).toMatchInlineSnapshot(`
    {
      "names": [
        "Infinity",
        "NaN",
        "undefined",
        "globalThis",
        "eval",
        "isFinite",
        "isNaN",
        "parseFloat",
        "parseInt",
        "decodeURI",
        "decodeURIComponent",
        "encodeURI",
        "encodeURIComponent",
        "AggregateError",
        "Array",
        "ArrayBuffer",
        "Boolean",
        "BigInt",
        "BigInt64Array",
        "BigUint64Array",
        "DataView",
        "Date",
        "Error",
        "EvalError",
        "FinalizationRegistry",
        "Float32Array",
        "Float64Array",
        "Function",
        "Int8Array",
        "Int16Array",
        "Int32Array",
        "Map",
        "Number",
        "Object",
        "Promise",
        "Proxy",
        "RangeError",
        "ReferenceError",
        "RegExp",
        "Set",
        "String",
        "Symbol",
        "SyntaxError",
        "TypeError",
        "Uint8Array",
        "Uint8ClampedArray",
        "Uint16Array",
        "Uint32Array",
        "URIError",
        "WeakMap",
        "WeakRef",
        "WeakSet",
        "JSON",
        "Math",
        "Reflect",
      ],
    }
  `);
});

test('call function on', async () => {
  const agent = new Agent();
  setSurroundingAgent(agent);
  const inspector = new TestInspector();
  const realm = new ManagedRealm();
  inspector.attachAgent(agent, [realm]);

  inspector.flush();
  await inspector.eval('const a = { x: 1 }; a');
  expect(inspector.flush()).toMatchInlineSnapshot(`
    [
      {
        "method": "Debugger.scriptParsed",
        "params": {
          "buildId": "",
          "endColumn": 21,
          "endLine": 1,
          "executionContextId": 0,
          "hash": "",
          "isModule": false,
          "scriptId": "0",
          "startColumn": 0,
          "startLine": 0,
          "url": "vm:///0",
        },
      },
      {
        "id": 0,
        "method": "Runtime.evaluate",
        "params": {
          "expression": "const a = { x: 1 }; a",
          "uniqueContextId": "0",
        },
      },
      {
        "id": 0,
        "result": {
          "exceptionDetails": undefined,
          "result": {
            "className": "Object",
            "description": "Object",
            "objectId": "default:1",
            "preview": {
              "description": "Object",
              "entries": undefined,
              "overflow": false,
              "properties": [
                {
                  "name": "x",
                  "type": "number",
                  "value": "1",
                },
              ],
              "subtype": undefined,
              "type": "object",
            },
            "subtype": undefined,
            "type": "object",
          },
        },
      },
    ]
  `);
  expect(await inspector.runtime.callFunctionOn({
    functionDeclaration: 'function (x) { return this.x + x }',
    arguments: [{ value: 1 }],
    executionContextId: 0,
    objectId: 'default:1',
  })).toMatchInlineSnapshot(`
    {
      "description": "2",
      "type": "number",
      "value": 2,
    }
  `);
  expect(await inspector.runtime.callFunctionOn({
    functionDeclaration: 'function (x) { return [this.x, x, 2, 3] }',
    arguments: [{ value: 1 }],
    executionContextId: 0,
    objectId: 'default:1',
    returnByValue: true,
  })).toMatchInlineSnapshot(`
    {
      "type": "object",
      "value": [
        1,
        1,
        2,
        3,
      ],
    }
  `);
});
