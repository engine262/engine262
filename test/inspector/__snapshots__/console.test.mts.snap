// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preview evaluation > [] 1`] = `
{
  "className": "Array",
  "description": "Array(0)",
  "objectId": "default:1",
  "preview": {
    "description": "Array(0)",
    "overflow": false,
    "properties": [],
    "subtype": "array",
    "type": "object",
  },
  "subtype": "array",
  "type": "object",
}
`;

exports[`preview evaluation > [1, 2, 3].map(x => x + 1) 1`] = `
{
  "className": "Array",
  "description": "Array(3)",
  "objectId": "default:4",
  "preview": {
    "description": "Array(3)",
    "overflow": false,
    "properties": [
      {
        "name": "0",
        "type": "number",
        "value": "2",
      },
      {
        "name": "1",
        "type": "number",
        "value": "3",
      },
      {
        "name": "2",
        "type": "number",
        "value": "4",
      },
    ],
    "subtype": "array",
    "type": "object",
  },
  "subtype": "array",
  "type": "object",
}
`;

exports[`preview evaluation > { let a = 1; a } 1`] = `
{
  "description": "1",
  "type": "number",
  "value": 1,
}
`;

exports[`preview evaluation > { var a = 1; a } 1`] = `"side effect"`;

exports[`preview evaluation > 1 1`] = `
{
  "description": "1",
  "type": "number",
  "value": 1,
}
`;

exports[`preview evaluation > globalThis.x = 1 1`] = `"side effect"`;
